# Analyzing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Keinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [DataQuest](https://www.dataquest.io/) including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- ``dateCrawled`` - When this ad was first crawled. All field-values are taken from this date.
- ``name`` - Name of the car.
- ``seller`` - Whether the seller is private or a dealer.
- ``offerType`` - The type of listing
- ``price`` - The price on the ad to sell the car.
- ``abtest`` - Whether the listing is included in an A/B test.
- ``vehicleType`` - The vehicle Type.
- ``yearOfRegistration`` - The year in which which year the car was first registered.
- ``gearbox`` - The transmission type.
- ``powerPS`` - The power of the car in PS.
- ``model`` - The car model name.
- ``kilometer`` - How many kilometers the car has driven.
- ``monthOfRegistration`` - The month in which which year the car was first registered.
- ``fuelType`` - What type of fuel the car uses.
- ``brand`` - The brand of the car.
- ``notRepairedDamage`` - If the car has a damage which is not yet repaired.
- ``dateCreated`` - The date on which the eBay listing was created.
- ``nrOfPictures`` - The number of pictures in the ad.
- ``postalCode`` - The postal code for the location of the vehicle.
- ``lastSeenOnline`` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [38]:
import pandas as pd
import numpy as np

In [39]:
autos = pd.read_csv(r"D:\autos.csv", encoding="Latin-1")
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

Our dataset contains 20 columns, most of wich are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

# Clean Columns

In [40]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

- Change the columns from camelcase to snakecase.
- Change a few wordings to more accurately describe the columns.

In [41]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

          date_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

# Initial Data Exploration and Cleaning

We'll start by exploring the data to find obvious areas where we can clean the data.

In [42]:
autos.describe(include="all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage         date_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

Our initial observation:

- There are a number of text columns where all (or nearly all) of the values are the same:
    - `seller`
    - `offer_type`
- The `nr_of_pictures` column looks odd, we'll need to investigate this further.

In [43]:
print(autos.nr_of_pictures.value_counts())
print("\n")
print(autos.offer_type.value_counts())
print("\n")
print(autos.seller.value_counts())

0    371528
Name: nr_of_pictures, dtype: int64


Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64


privat        371525
gewerblich         3
Name: seller, dtype: int64


It looks like the `nr_of_pictures` column has`0` for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [44]:
autos.drop(["seller", "offer_type", "nr_of_pictures"], inplace=True, axis=1)

# Exploring Price and Kilometer

In [45]:
autos.kilometer.value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: kilometer, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [46]:
print(autos.price.unique().shape)
print(autos.price.describe())
autos.price.value_counts().head(20)

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

Again, the prices seem rounded, however given there are 5597 unique values in the column, that may just be people's tendency to round prices on the site.

There are 10778 cars listed with ``$0`` price - given that this is only 3% of the cars, we might consider removing these rows. The maximum price is ``$2,147,483,647.00``, which seems a lot, let's look at the highest prices further.

In [47]:
autos.price.value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

In [48]:
autos.price.value_counts().sort_index(ascending=False).head(60)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
1111111        2
1010010        1
1000000        5
999999        13
999990         1
911911         1
849000         1
820000         1
780000         1
745000         2
725000         1
700000         1
650000         1
619000         1
600000         2
599000         1
585000         1
579000         1
517895         1
500000         2
488997         1
487000         1
485000         1
466000         1
445000         1
440000         1
420000         1
399997         1
395000         1
390000         1
370000         2
368000         1
350000         4
349000         1
345000         1
323223         1
300000         1
299000         3
295000        

There area a number of listings below `$20`, including 10778 at `$0`. There are also a small number of listings with very high values, including 52 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is ``$1``. We will keep the ``$1`` items, but remove anything above $849,000 since it seems that prices increase steadely to that number.

In [49]:
autos = autos[autos.price.between(1,849000)]
autos.price.describe()

count    360665.000000
mean       5944.315223
std       10262.634956
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      849000.000000
Name: price, dtype: float64

# Exploring the Date Columns

There are a number of columns with date information:

- `date_crawled`
- `date_created`
- `registration_month`
- `registration_year`
- `last_seen`

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [50]:
autos[['date_crawled', 'date_created', 'last_seen']].head()

date_crawled         date_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [51]:
(
    autos.date_crawled.str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
)

2016-03-05    0.025547
2016-03-06    0.014482
2016-03-07    0.035654
2016-03-08    0.033469
2016-03-09    0.034112
2016-03-10    0.032648
2016-03-11    0.032770
2016-03-12    0.036244
2016-03-13    0.015782
2016-03-14    0.036327
2016-03-15    0.033422
2016-03-16    0.030203
2016-03-17    0.031650
2016-03-18    0.013117
2016-03-19    0.035271
2016-03-20    0.036399
2016-03-21    0.035690
2016-03-22    0.032493
2016-03-23    0.032005
2016-03-24    0.029914
2016-03-25    0.032798
2016-03-26    0.031971
2016-03-27    0.030230
2016-03-28    0.035060
2016-03-29    0.034123
2016-03-30    0.033533
2016-03-31    0.031877
2016-04-01    0.034148
2016-04-02    0.035094
2016-04-03    0.038812
2016-04-04    0.037625
2016-04-05    0.012785
2016-04-06    0.003128
2016-04-07    0.001619
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [52]:
(
    autos.last_seen.str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
)

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007938
2016-03-09    0.009824
2016-03-10    0.011459
2016-03-11    0.012954
2016-03-12    0.023243
2016-03-13    0.008409
2016-03-14    0.012175
2016-03-15    0.016323
2016-03-16    0.016417
2016-03-17    0.028703
2016-03-18    0.006887
2016-03-19    0.016331
2016-03-20    0.019888
2016-03-21    0.020030
2016-03-22    0.020507
2016-03-23    0.018019
2016-03-24    0.019162
2016-03-25    0.018998
2016-03-26    0.015957
2016-03-27    0.016719
2016-03-28    0.022187
2016-03-29    0.023285
2016-03-30    0.023723
2016-03-31    0.024241
2016-04-01    0.023895
2016-04-02    0.024965
2016-04-03    0.025320
2016-04-04    0.025533
2016-04-05    0.126963
2016-04-06    0.218948
2016-04-07    0.130434
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of "last seen" values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [53]:
print(autos.date_created.str[:10].unique().shape)

(
    autos.date_created.str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
)

(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039000
2016-04-04    0.037733
2016-04-05    0.011617
2016-04-06    0.003119
2016-04-07    0.001555
Name: date_created, Length: 114, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 2 years.

In [54]:
autos.registration_year.describe()

count    360665.000000
mean       2004.432817
std          81.013776
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is `1000`, long before cars were invented and the maximum is `9999`, many years into the future.

# Dealing with Incorrect Registration Year Data


Because a car can't be registered before the listing was seen, any vehicle with a registration year above 2019 is definitely innacurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [55]:
(~autos.registration_year.between(1900,2019)).sum() / autos.registration_year.shape[0]

0.0004103530977499896

Given that this is less than 0.1% of our data, we will remove theses rows.

In [56]:
autos = autos[autos.registration_year.between(1900,2019)]
autos.registration_year.value_counts(normalize=True).head(10)

2000    0.064136
1999    0.061109
2005    0.060263
2006    0.055490
2001    0.054769
2003    0.054383
2004    0.054020
2002    0.052203
2007    0.048558
1998    0.047784
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

# Exploring Price by Brand

In [57]:
autos.brand.value_counts(normalize=True)

volkswagen        0.213541
bmw               0.108580
opel              0.107379
mercedes_benz     0.095987
audi              0.088709
ford              0.068904
renault           0.048336
peugeot           0.029979
fiat              0.025960
seat              0.018948
skoda             0.015464
mazda             0.015367
smart             0.014352
citroen           0.014033
nissan            0.013600
toyota            0.012826
hyundai           0.009961
sonstige_autos    0.009459
mini              0.009312
volvo             0.008982
mitsubishi        0.008219
honda             0.007584
kia               0.006926
suzuki            0.006316
alfa_romeo        0.006274
porsche           0.006055
chevrolet         0.004971
chrysler          0.003828
dacia             0.002469
jeep              0.002180
daihatsu          0.002150
land_rover        0.002103
subaru            0.002080
jaguar            0.001689
daewoo            0.001478
saab              0.001440
trabant           0.001376
l

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [58]:
brand_counts = autos.brand.value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [59]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos.brand == brand]
    mean_price = brand_only.price.mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5333,
 'bmw': 8400,
 'opel': 2940,
 'mercedes_benz': 8490,
 'audi': 8979,
 'ford': 3660}

Of the top 5 brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

# Exploring Mileage

In [60]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=['mean_price'])

mean_price
volkswagen           5333
bmw                  8400
opel                 2940
mercedes_benz        8490
audi                 8979
ford                 3660

In [61]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos.brand == brand]
    mean_mileage = brand_only.kilometer.mean()
    brand_mean_mileage[brand] = mean_mileage
    
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)

In [63]:
brand_info = pd.DataFrame(mean_prices, columns=["mean_prices"])
brand_info['mean_mileage'] = mean_mileage
brand_info

mean_prices   mean_mileage
audi                  8979  129692.160971
mercedes_benz         8490  130703.511053
bmw                   8400  132949.163367
volkswagen            5333  128662.661557
ford                  3660  123915.100036
opel                  2940  128930.305848

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.